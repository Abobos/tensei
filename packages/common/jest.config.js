module.exports = {
    testEnvironment: 'node',
    testPathIgnorePatterns: [
        '/node_modules/',
        '<rootDir>/src/__tests__/helpers',
        '<rootDir>/build'
    ]
}
