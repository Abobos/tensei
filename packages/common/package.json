{
    "name": "@tensei/common",
    "version": "0.5.2",
    "main": "./build/index.js",
    "license": "MIT",
    "types": "./typings/index.d.ts",
    "files": [
        "build/",
        "typings/"
    ],
    "devDependencies": {
        "@babel/core": "^7.11.4",
        "@babel/preset-env": "^7.11.0",
        "@babel/preset-typescript": "^7.10.4",
        "@types/faker": "^4.1.12",
        "babel-jest": "^26.3.0",
        "faker": "^5.1.0",
        "jest": "^26.0.1",
        "prettier": "^2.0.5",
        "supertest": "^4.0.2",
        "typescript": "^3.9.7"
    },
    "scripts": {
        "build": "tsc --p tsconfig.json",
        "dev": "tsc --watch --p tsconfig.json",
        "prettier": "prettier --write './**/*.{js,json,ts,css}'",
        "test": "jest --verbose --runInBand --forceExit"
    },
    "dependencies": {
        "@tensei/mail": "^0.5.1",
        "@types/luxon": "^1.25.0",
        "@types/pluralize": "^0.0.29",
        "@types/signale": "^1.4.1",
        "change-case": "^4.1.1",
        "csstype": "^3.0.2",
        "indicative": "^7.4.4",
        "luxon": "^1.25.0",
        "pluralize": "^8.0.0",
        "signale": "^1.4.0"
    },
    "publishConfig": {
        "access": "public"
    },
    "gitHead": "e891d0776360c97f7d8035cb283b296739c29ff8"
}
