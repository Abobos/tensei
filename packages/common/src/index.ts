import Resource from './resources/Resource'

export { text, Text } from './fields/Text'
export { link, Link } from './fields/Link'
export { date, DateField } from './fields/Date'
export { hasOne, HasOne } from './fields/HasOne'
export { select, Select } from './fields/Select'
export { hasMany, HasMany } from './fields/HasMany'
export { integer, Integer } from './fields/Integer'
export { number, NumberField } from './fields/Number'
export { dateTime, DateTime } from './fields/DateTime'
export { password, Password } from './fields/Password'
export { textarea, Textarea } from './fields/Textarea'
export { Field, SerializedField } from './fields/Field'
export { belongsTo, BelongsTo } from './fields/BelongsTo'
export { bigInteger, BigInteger } from './fields/BigInteger'
export { belongsToMany, BelongsToMany } from './fields/BelongsToMany'

export { Tool, tool, ToolSetupConfig } from './tools/Tool'
export { ResourceManager } from './resources/ResourceManager'
export { action, SerializedAction, Action } from './actions/Action'
export { resource, Resource, SerializedResource } from './resources/Resource'
export { DatabaseRepositoryInterface } from './databases/DatabaseRepositoryInterface'

export * from './config'
