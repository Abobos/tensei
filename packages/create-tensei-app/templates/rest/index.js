const { cms } = require('@tensei/cms')
const { auth } = require('@tensei/auth')
const { rest } = require('@tensei/rest')
const { tensei } = require('@tensei/core')
const { media } = require('@tensei/media')

tensei()
    .root(__dirname)
    .plugins([
        cms().plugin(),
        media().plugin(),
        auth().jwt().plugin(),
        rest().plugin()
    ])
    .start()
    .catch(console.error)
