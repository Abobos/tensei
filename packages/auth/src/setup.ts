import { sentenceCase } from 'change-case'
import { EntityManager } from '@mikro-orm/core'
import { ResourceContract, Permission, PluginSetupConfig } from '@tensei/common'

import { AuthPluginConfig } from './config'

export const setup = async (
    config: PluginSetupConfig,
    authConfig: AuthPluginConfig
) => {
    const { resources, orm } = config

    const { em } = orm!
    const RoleResource = authConfig.resources.role
    const PermissionResource = authConfig.resources.permission

    const insertPermissions = await getPermissionsToInsert(
        resources,
        em,
        PermissionResource
    )

    if (insertPermissions.length > 0) {
        await em.persistAndFlush(
            insertPermissions.map(permission =>
                em.create(PermissionResource.data.pascalCaseName, permission)
            )
        )
    }

    // Insert default roles
    // There will be two default roles: Authenticated & Public
    // Public will have no permissions attached by default
    let [authenticatedRole, publicRole] = await em.find(
        RoleResource.data.pascalCaseName,
        {
            slug: {
                $in: ['authenticated', 'public']
            }
        },
        {
            orderBy: {
                slug: 'ASC'
            }
        }
    )

    const rolesToCreate = []
    const allPermissions = (
        await em.find(PermissionResource.data.pascalCaseName, {})
    ).map((permission: any) => permission.id)

    if (!authenticatedRole) {
        rolesToCreate.push({
            name: 'Authenticated',
            slug: 'authenticated',
            permissions: allPermissions
        })
    }

    if (!publicRole) {
        rolesToCreate.push({
            name: 'Public',
            slug: 'public',
            permissions: allPermissions
        })
    }

    await em.persistAndFlush(
        rolesToCreate.map(role =>
            em.create(RoleResource.data.pascalCaseName, role)
        )
    )
}

const getPermissionsToInsert = async (
    resources: ResourceContract[],
    em: EntityManager,
    permissionResource: ResourceContract
) => {
    const permissions: Permission[] = []

    resources.forEach(resource => {
        ;['insert', 'fetch', 'show', 'update', 'delete'].forEach(operation => {
            permissions.push(`${operation}:${resource.data.slug}`)
        })

        resource.data.actions.forEach(action => {
            permissions.push(`run:${resource.data.slug}:${action.data.slug}`)
        })

        resource.data.permissions.forEach(permission => {
            permissions.push(permission)
        })
    })

    const existingPermissions: any[] = (
        await em.find(permissionResource.data.pascalCaseName, {
            slug: {
                $in: permissions
            }
        })
    ).map(permission => permission.slug)

    const newPermissionsToCreate = permissions.filter(
        permission =>
            !existingPermissions.includes(
                typeof permission === 'string' ? permission : permission.slug
            )
    )

    return Array.from(new Set(newPermissionsToCreate)).map(permission => ({
        name:
            typeof permission === 'string'
                ? sentenceCase(permission.split(':').join(' '))
                : permission.name,
        slug: permission
    }))
}
