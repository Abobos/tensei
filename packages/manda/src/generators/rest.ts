import { PluginSetupConfig, ResourceContract } from '@tensei/common'
import { PluginContract } from '@tensei/common/plugins'
import { generateAuthApi } from './auth'

const generateApiClassForResource = (resource: ResourceContract, path: string) => {
	if (resource.data.hideOnApi) {
		return ``
	}

	const apiPrefix = `${path}/`

	return `
        export class ${resource.data.pascalCaseName}API {
            constructor(private options: SdkOptions, private instance: AxiosInstance) {}

            ${
							resource.data.hideOnFetchApi
								? ''
								: `
            /**
             * 
             * Fetch a single ${resource.data.pascalCaseName.toLowerCase()} from the API.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}().find({ id })
             * 
             **/
            find(payload: {
                id: Tensei.${resource.data.pascalCaseName}['id'],
                select?: Tensei.${resource.data.pascalCaseName}SelectFields[],
                populate?: Tensei.${resource.data.pascalCaseName}PopulateFields[],
            }) {
                return this.instance.get<Tensei.FindResponse<Tensei.${
									resource.data.pascalCaseName
								}>>('${apiPrefix}' + '${resource.data.slugPlural}/' + payload.id)
            }

            /**
             * 
             * Fetch a paginated list of ${resource.data.camelCaseNamePlural} from the API.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.findMany({
             *          where: { id: { _in: [1, 2] } },
             *          sort: { id: SortQueryInput.ASC },
             *          pagination: { per_page: 30, page: 1 },
             *      })
             *
             **/
            findMany(payload: {
                where?: Tensei.${resource.data.pascalCaseName}WhereQueryInput,
                sort?: Tensei.${resource.data.pascalCaseName}SortQueryInput,
                pagination?: Tensei.PaginationOptions,
                fields?: Tensei.${resource.data.pascalCaseName}SelectFields[],
                populate?: Tensei.${resource.data.pascalCaseName}PopulateFields[]
            } = {}) {
                return this.instance.get<Tensei.PaginatedResponse<Tensei.${
									resource.data.pascalCaseName
								}>>('${apiPrefix}' + '${resource.data.slugPlural}', {
                                    params: {
                                        populate: payload?.populate?.join(',') || [],
                                        per_page: payload?.pagination?.per_page,
                                        page: payload?.pagination?.page,
                                        fields: payload?.fields?.join(',') || undefined,
                                        where: payload?.where
                                    }
                                })
            }
            `
						}

            ${
							resource.data.hideOnInsertApi
								? ''
								: `
            /**
             * 
             * Insert a single ${resource.data.pascalCaseName.toLowerCase()}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.insert({ object: {...} })
             *
             **/
            insert(payload: {
                object: Tensei.${resource.data.pascalCaseName}InsertInput
            }) {
                return this.instance.post<Tensei.FindResponse<Tensei.${
									resource.data.pascalCaseName
								}>>('${apiPrefix}' + '${resource.data.slugPlural}', payload.object)
            }

            /**
             * 
             * Insert multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${
								resource.data.camelCaseNamePlural
							}.insertMany({ objects: [{...}, {...}] })
             *
             **/
            insertMany(payload: {
                objects: Tensei.${resource.data.pascalCaseName}InsertInput[]
            }) {
                return this.instance.post<Tensei.FindResponse<Tensei.${
									resource.data.pascalCaseName
								}[]>>('${apiPrefix}' + '${resource.data.slugPlural}/bulk', payload)
            }
            `
						}

            ${
							resource.data.hideOnUpdateApi
								? ''
								: `
            /**
             * 
             * Update a single ${resource.data.pascalCaseName.toLowerCase()}.
             *    Example:
             *      await tensei.${
								resource.data.camelCaseNamePlural
							}.update({ id: 1, object: {...} })
             *
             **/
            update(payload: {
                id: Tensei.${resource.data.pascalCaseName}['id']
                object: Tensei.${resource.data.pascalCaseName}UpdateInput
            }) {
                return this.instance.patch<Tensei.FindResponse<Tensei.${
									resource.data.pascalCaseName
								}>>('${apiPrefix}' + '${resource.data.slugPlural}/' + payload.id, payload.object)
            }

            /**
             * 
             * Update multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.updateMany({
             *          where: { id: { _in: [1, 2] } },
             *          object: {...},
             *      })
             *
             **/
            updateMany(payload: {
                object: Tensei.${resource.data.pascalCaseName}UpdateInput,
                where: Tensei.${resource.data.pascalCaseName}WhereQueryInput
            }) {
                return this.instance.patch('${apiPrefix}' + '${
										resource.data.slugPlural
								  }/bulk', payload) as Promise<Tensei.FindResponse<Tensei.${
										resource.data.pascalCaseName
								  }[]>>
            }
            `
						}

            ${
							resource.data.hideOnDeleteApi
								? ''
								: `
            /**
             * 
             * Delete single ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.delete({
             *          id: 1
             *      })
             *
             **/
            delete(payload: {
                id: Tensei.${resource.data.pascalCaseName}['id']
            }) {
                return this.instance.delete('${apiPrefix}' + '${resource.data.slugPlural}' + payload.id) as Promise<Tensei.FindResponse<Tensei.${resource.data.pascalCaseName}>>
            }

            /**
             * 
             * Delete multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.deleteMany({
             *          where: { id: { _in: [1, 2] } },
             *      })
             *
             **/
            deleteMany(payload: {
                where: Tensei.${resource.data.pascalCaseName}WhereQueryInput
            }) {
                return this.instance.delete('${apiPrefix}' + '${resource.data.slugPlural}', {
                    params: {
                        where: payload.where
                    }
                }) as Promise<Tensei.FindResponse<Tensei.${resource.data.pascalCaseName}[]>>
            }
            `
						}
        }
    `
}

const generateAPIClient = (resources: ResourceContract[], plugins: PluginContract[]) => {
	const authPlugin = plugins.find((plugin) => plugin.config.name === 'Auth')

	return `
    export interface SdkOptions {
        url?: string
        axiosInstance?: AxiosInstance
        axiosRequestConfig?: Omit<AxiosRequestConfig, 'baseURL'>
    }

    export class Sdk {
        private instance: AxiosInstance
        ${resources
					.filter((resource) => !resource.data.hideOnApi)
					.map(
						(resource) =>
							`public ${resource.data.camelCaseNamePlural} : ${resource.data.pascalCaseName}API\n`
					)
					.join('')}

        ${authPlugin ? 'public auth: AuthAPI\n' : ''}

        constructor(private options?: SdkOptions) {
            this.instance = options?.axiosInstance || Axios.create({
                baseURL: this.options?.url || 'http://localhost:8810',
                ...(options?.axiosRequestConfig || {})
            })

            ${authPlugin ? 'this.auth = new AuthAPI(this.instance)' : ''}
            ${resources
							.filter((resource) => !resource.data.hideOnApi)
							.map((resource) => {
								return `this.${resource.data.camelCaseNamePlural} =  new ${resource.data.pascalCaseName}API(this.options, this.instance)
        `
							})
							.join('')}
        }
    }
`
}

const generateImports = () => {
	return `
    import * as Tensei from './interfaces'
    import Axios, { AxiosInstance, AxiosRequestConfig } from 'axios'

    export * from './interfaces'
    `
}

export const generateFetchWrapperForResources = (config: PluginSetupConfig) => {
	const { resources, plugins } = config

	const restPlugin = plugins.find((plugin) => plugin.config.name === 'Rest API')

	if (!restPlugin) {
		return ``
	}

	const path = restPlugin.config.extra?.path || 'api'

	return (
		[generateImports(), generateAPIClient(resources, plugins), generateAuthApi(config)]
			.concat(
				resources.map(
					(resource) => `
        ${generateApiClassForResource(resource, path)}
    `
				)
			)
			// .concat([generateAPIClient(resources)])
			.join('')
	)
}
