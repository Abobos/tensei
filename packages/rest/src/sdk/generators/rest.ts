import { PluginSetupConfig, ResourceContract } from '@tensei/common'
import { PluginContract } from '@tensei/common/plugins'
import { generateAuthApi } from './auth'

const generateApiClassForResource = (
  resource: ResourceContract,
  path: string
) => {
  if (resource.data.hideOnApi) {
    return ``
  }

  const apiPrefix = `${path}/`

  return `
        export interface ${resource.data.pascalCaseName}SdkContract {
            ${
              resource.data.hideOnFetchApi
                ? ''
                : `
            /**
             * 
             * Fetch a single ${resource.data.pascalCaseName.toLowerCase()} from the API.
             *    Example:
             *      await tensei.${
               resource.data.camelCaseNamePlural
             }().find({ id })
             * 
             **/
            find(payload: {
                id: ${resource.data.pascalCaseName}['id'],
                select?: ${resource.data.pascalCaseName}SelectFields[],
                populate?: ${resource.data.pascalCaseName}PopulateFields[],
            }): Promise<FindResponse<${resource.data.pascalCaseName}>>

            /**
             * 
             * Fetch a paginated list of ${
               resource.data.camelCaseNamePlural
             } from the API.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.findMany({
             *          where: { id: { _in: [1, 2] } },
             *          sort: { id: SortQueryInput.ASC },
             *          pagination: { per_page: 30, page: 1 },
             *      })
             *
             **/
            findMany(payload?: {
                where?: ${resource.data.pascalCaseName}WhereQueryInput,
                sort?: ${resource.data.pascalCaseName}SortQueryInput,
                pagination?: PaginationOptions,
                fields?: ${resource.data.pascalCaseName}SelectFields[],
                populate?: ${resource.data.pascalCaseName}PopulateFields[]
            }): Promise<PaginatedResponse<${resource.data.pascalCaseName}>>
            `
            }

            ${
              resource.data.hideOnCreateApi
                ? ''
                : `
            /**
             * 
             * Create a single ${resource.data.pascalCaseName.toLowerCase()}.
             *    Example:
             *      await tensei.${
               resource.data.camelCaseNamePlural
             }.create({ object: {...} })
             *
             **/
            create(payload: {
                object: ${resource.data.pascalCaseName}CreateInput
            }): Promise<FindResponse<${resource.data.pascalCaseName}>>

            /**
             * 
             * Create multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${
               resource.data.camelCaseNamePlural
             }.createMany({ objects: [{...}, {...}] })
             *
             **/
            createMany(payload: {
                objects: ${resource.data.pascalCaseName}CreateInput[]
            }): Promise<FindResponse<${resource.data.pascalCaseName}[]>>
            `
            }

            ${
              resource.data.hideOnUpdateApi
                ? ''
                : `
            /**
             * 
             * Update a single ${resource.data.pascalCaseName.toLowerCase()}.
             *    Example:
             *      await tensei.${
               resource.data.camelCaseNamePlural
             }.update({ id: 1, object: {...} })
             *
             **/
            update(payload: {
                id: ${resource.data.pascalCaseName}['id']
                object: ${resource.data.pascalCaseName}UpdateInput
            }): Promise<FindResponse<${resource.data.pascalCaseName}>>

            /**
             * 
             * Update multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${
               resource.data.camelCaseNamePlural
             }.updateMany({
             *          where: { id: { _in: [1, 2] } },
             *          object: {...},
             *      })
             *
             **/
            updateMany(payload: {
                object: ${resource.data.pascalCaseName}UpdateInput,
                where: ${resource.data.pascalCaseName}WhereQueryInput
            }): Promise<FindResponse<${resource.data.pascalCaseName}[]>>
            `
            }

            ${
              resource.data.hideOnDeleteApi
                ? ''
                : `
            /**
             * 
             * Delete single ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.delete({
             *          id: 1
             *      })
             *
             **/
            delete(payload: {
                id: ${resource.data.pascalCaseName}['id']
            }): Promise<FindResponse<${resource.data.pascalCaseName}>>

            /**
             * 
             * Delete multiple ${resource.data.camelCaseNamePlural}.
             *    Example:
             *      await tensei.${resource.data.camelCaseNamePlural}.deleteMany({
             *          where: { id: { _in: [1, 2] } },
             *      })
             *
             **/
            deleteMany(payload: {
                where: ${resource.data.pascalCaseName}WhereQueryInput
            }): Promise<FindResponse<${resource.data.pascalCaseName}[]>>
            `
            }
        }
    `
}

const generateAPIClient = (
  resources: ResourceContract[],
  plugins: PluginContract[]
) => {
  const authPlugin = plugins.find(plugin => plugin.config.name === 'Auth')

  return `
    export interface SdkOptions {
        url?: string
        refreshTokens?: boolean
        axiosInstance?: AxiosInstance
        axiosRequestConfig?: Omit<AxiosRequestConfig, 'baseURL'>
    }

    export interface SdkContract {
        ${resources
          .filter(resource => !resource.data.hideOnApi)
          .map(
            resource =>
              `${resource.data.camelCaseNamePlural}() : ${resource.data.pascalCaseName}SdkContract\n`
          )
          .join('')}

        ${authPlugin ? 'auth(): AuthSdkContract\n' : ''}
        options?: SdkOptions
    }

    export const sdk: (options?: SdkOptions) => SdkContract
`
}

const generateImports = () => {
  return `
    import { AxiosInstance, AxiosRequestConfig } from 'axios'
    `
}

export const generateFetchWrapperForResources = (config: PluginSetupConfig) => {
  const { resources, plugins } = config

  const restPlugin = plugins.find(plugin => plugin.config.name === 'Rest API')

  if (!restPlugin) {
    return ``
  }

  const path = restPlugin.config.extra?.path || 'api'

  return (
    [
      generateImports(),
      generateAPIClient(resources, plugins),
      generateAuthApi(config)
    ]
      .concat(
        resources.map(
          resource => `
        ${generateApiClassForResource(resource, path)}
    `
        )
      )
      // .concat([generateAPIClient(resources)])
      .join('')
  )
}
