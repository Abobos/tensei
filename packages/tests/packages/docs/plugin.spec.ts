import { setup } from './setup'
import Supertest from 'supertest'
import { rest } from '@tensei/rest'
import { route, plugin } from '@tensei/common'

test('Generates a new tag for each resource exposed to the API', async () => {
    const { app } = await setup()

    const client = Supertest(app)
    const response = await client.get('/docs.json')

    expect(response.body.tags).toEqual([
        { name: 'Tags' },
        { name: 'Comments' },
        { name: 'Users' },
        { name: 'Posts' },
        { name: 'Reactions' }
    ])
})

test('Generates a new object definition for all resources exposed to the API', async () => {
    const { app } = await setup()

    const client = Supertest(app)
    const response = await client.get('/docs.json')

    expect(response.body.definitions.ID).toEqual({
        type: 'string'
    })
    expect(response.body.definitions.PaginationMeta).toBeDefined()

    expect(response.body.definitions.Tag).toBeDefined()
    expect(response.body.definitions.TagInput).toBeDefined()
    expect(response.body.definitions.TagFetchResponse).toBeDefined()

    expect(response.body.definitions.Comment).toBeDefined()
    expect(response.body.definitions.CommentInput).toBeDefined()
    expect(response.body.definitions.CommentFetchResponse).toBeDefined()

    expect(response.body.definitions.User).toBeDefined()
    expect(response.body.definitions.UserInput).toBeDefined()
    expect(response.body.definitions.UserFetchResponse).toBeDefined()

    expect(response.body.definitions.Post).toBeDefined()
    expect(response.body.definitions.PostInput).toBeDefined()
    expect(response.body.definitions.PostFetchResponse).toBeDefined()

    expect(response.body.definitions.Reaction).toBeDefined()
    expect(response.body.definitions.ReactionInput).toBeDefined()
    expect(response.body.definitions.ReactionFetchResponse).toBeDefined()
})

test('Adds new route definitions to API documentation', async () => {
    const path = '/accessories/query'

    const { app } = await setup([
        plugin('New route').boot(async ({ extendRoutes }) => {
            extendRoutes([
                route('Fetch Accessories')
                    .get()
                    .path(path)
                    .extend({
                        docs: {
                            parameters: [
                                {
                                    required: true,
                                    name: 'id',
                                    in: 'body'
                                }
                            ],
                            tags: ['Accessory'],
                            responses: {
                                200: {
                                    description:
                                        'Accessories found successfully.'
                                }
                            }
                        }
                    })
            ])
        })
    ])

    const client = Supertest(app)
    const response = await client.get('/docs.json')

    expect(response.body.paths[path]).toEqual({
        get: {
            consumes: ['application/json'],
            produces: ['application/json'],
            tags: ['Accessory'],
            parameters: [
                {
                    in: 'body',
                    name: 'id',
                    required: true
                }
            ],
            responses: {
                500: {
                    description: expect.any(String)
                },
                200: {
                    description: 'Accessories found successfully.'
                }
            }
        }
    })
})

test('Generates all routes for REST api', async () => {
    const { app } = await setup()

    const client = Supertest(app)
    const response = await client.get('/docs.json')

    expect(response.body).toMatchSnapshot()
})
