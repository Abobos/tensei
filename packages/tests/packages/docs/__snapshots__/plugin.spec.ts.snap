// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates all routes for REST api 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "bearerAuth": Object {
        "bearerFormat": "JWT",
        "scheme": "bearer",
      },
    },
  },
  "definitions": Object {
    "Comment": Object {
      "properties": Object {},
      "type": "object",
    },
    "CommentFetchResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "$ref": "#/definitions/Comment",
          },
          "type": "array",
        },
        "meta": Object {
          "$ref": "#/definitions/PaginationMeta",
          "type": "object",
        },
      },
      "type": "object",
    },
    "CommentInput": Object {
      "properties": Object {
        "body": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "post": Object {
          "$ref": "#/definitions/ID",
          "description": "",
        },
        "title": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "title",
        "body",
        "post",
      ],
      "type": "object",
    },
    "ID": Object {
      "type": "string",
    },
    "PaginationMeta": Object {
      "properties": Object {
        "page": Object {
          "type": "integer",
        },
        "page_count": Object {
          "type": "integer",
        },
        "per_page": Object {
          "type": "integer",
        },
        "total": Object {
          "type": "integer",
        },
      },
      "type": "object",
    },
    "Post": Object {
      "properties": Object {},
      "type": "object",
    },
    "PostFetchResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "$ref": "#/definitions/Post",
          },
          "type": "array",
        },
        "meta": Object {
          "$ref": "#/definitions/PaginationMeta",
          "type": "object",
        },
      },
      "type": "object",
    },
    "PostInput": Object {
      "properties": Object {
        "approved": Object {
          "description": "",
          "type": "boolean",
        },
        "av_cpc": Object {
          "description": "",
          "format": "int32",
          "type": "integer",
        },
        "category": Object {
          "description": "",
          "enum": Array [
            "javascript",
            "angular",
            "mysql",
            "pg",
          ],
          "format": "string",
          "type": "string",
        },
        "content": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "description": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "published_at": Object {
          "description": "",
          "format": "date",
          "type": "string",
        },
        "scheduled_for": Object {
          "description": "",
          "format": "date",
          "type": "string",
        },
        "tags": Object {
          "description": "",
          "items": Object {
            "$ref": "#/definitions/ID",
          },
          "type": "array",
        },
        "title": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "user": Object {
          "$ref": "#/definitions/ID",
          "description": "",
        },
      },
      "required": Array [
        "id",
        "title",
        "description",
        "content",
        "av_cpc",
        "category",
        "user",
        "published_at",
        "approved",
        "scheduled_for",
      ],
      "type": "object",
    },
    "Reaction": Object {
      "properties": Object {},
      "type": "object",
    },
    "ReactionFetchResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "$ref": "#/definitions/Reaction",
          },
          "type": "array",
        },
        "meta": Object {
          "$ref": "#/definitions/PaginationMeta",
          "type": "object",
        },
      },
      "type": "object",
    },
    "ReactionInput": Object {
      "properties": Object {
        "like": Object {
          "description": "",
          "type": "boolean",
        },
      },
      "required": Array [
        "id",
        "like",
      ],
      "type": "object",
    },
    "Tag": Object {
      "properties": Object {},
      "type": "object",
    },
    "TagFetchResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "$ref": "#/definitions/Tag",
          },
          "type": "array",
        },
        "meta": Object {
          "$ref": "#/definitions/PaginationMeta",
          "type": "object",
        },
      },
      "type": "object",
    },
    "TagInput": Object {
      "properties": Object {
        "description": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "name": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "posts": Object {
          "description": "",
          "items": Object {
            "$ref": "#/definitions/ID",
          },
          "type": "array",
        },
      },
      "required": Array [
        "id",
        "name",
        "description",
      ],
      "type": "object",
    },
    "User": Object {
      "properties": Object {},
      "type": "object",
    },
    "UserFetchResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "$ref": "#/definitions/User",
          },
          "type": "array",
        },
        "meta": Object {
          "$ref": "#/definitions/PaginationMeta",
          "type": "object",
        },
      },
      "type": "object",
    },
    "UserInput": Object {
      "properties": Object {
        "email": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "full_name": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "password": Object {
          "description": "",
          "format": "string",
          "type": "string",
        },
        "posts": Object {
          "description": "",
          "items": Object {
            "$ref": "#/definitions/ID",
          },
          "type": "array",
        },
      },
      "required": Array [
        "id",
        "full_name",
        "email",
        "password",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "contact": Object {},
    "license": Object {},
  },
  "paths": Object {},
  "securityDefinitions": Object {
    "Bearer": Object {
      "in": "header",
      "name": "Authorization",
      "type": "apiKey",
    },
  },
  "swagger": "2.0",
  "tags": Array [
    Object {
      "name": "Tags",
    },
    Object {
      "name": "Comments",
    },
    Object {
      "name": "Users",
    },
    Object {
      "name": "Posts",
    },
    Object {
      "name": "Reactions",
    },
  ],
}
`;
